### PR TITLE
CMake fixes for mfem-cmake

### PR DESCRIPTION
**Pull request by joshessman-llnl**
See https://github.com/spack/spack/pull/22521 and https://github.com/LLNL/serac/pull/430

The order change of MFEM TPLs is required for static linking - dependents must always precede dependees and I ran into issues with hypre's dependency on BLAS and SuperLU's dependency on METIS.  There is probably more reshuffling required for static linking to work in the general case.

See https://github.com/LLNL/serac/pull/430#issuecomment-811250895 for an explanation of the HDF5 changes

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#2141](https://github.com/mfem/mfem/pull/2141) | @joshessman-llnl | @tzanio | @adrienbernede + @acfisher + @jandrej | 04/29/21 | 04/29/21 | 04/30/21 | |

### CODE REVIEW

**Comment by adrienbernede** [comment]
LGTM

**Comment by adrienbernede** [comment]
Thank you so much @joshessman-llnl for looking into this.
@tzanio @acfisher 
I think it would be optimal if those fixes could make it to 4.3. It just needs to be synced with similar changes there https://github.com/spack/spack/pull/23324. Looking into it right now.

**Comment by adrienbernede** [comment]
LGTM

**Comment by acfisher** [comment]
Let us know when this is out of WIP status and we can get it reviewed.

**Comment by adrienbernede** [comment]
@joshessman-llnl the issue in Gitlab CI will be solved merging current master in your branch.
This is due to an out-dated commit hash in `tests/gitlab/get_mfem_uberenv`.

File: config/cmake/modules/FindNetCDF.cmake [hunk 1]
```diff
@@ -15,5 +15,20 @@
...
+set(HDF5_FOUND OFF)
+enable_language(C) # FindHDF5.cmake uses the C compiler
 mfem_find_package(NetCDF NETCDF NETCDF_DIR "include" netcdf.h "lib" netcdf
   "Paths to headers required by NetCDF." "Libraries required by NetCDF.")
+# The HL extension libraries are in a separate variable and must precede
+# the "regular" hdf5 library, as hdf5_hl depends on hdf5
+# The netcdf library will always be the first element of NETCDF_LIBRARIES
+# and we need to insert after that library but before the hdf5 library, so
+# position 1 is used
+list(INSERT NETCDF_LIBRARIES 1 ${HDF5_C_LIBRARY_hdf5_hl})
```
Comment (adrienbernede) [inline config/cmake/modules/FindNetCDF.cmake:34]: Thanks for the helpful comments!

**Review by adrienbernede — APPROVED** [review]
Greatly appreciate your help on this @joshessman-llnl !

**Comment by adrienbernede** [comment]
LGTM

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Review by jandrej — APPROVED** [review]

**Review by acfisher — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Pull request merged**